ansi/buffer.go

```diff
@@ -11,28 +12,64 @@ type Buffer struct {
 	bytes.Buffer
 }
 
-// PrintableRuneWidth returns the width of all printable runes in the buffer.
+// PrintableRuneWidth returns the cell width of all printable runes in the
+// buffer.
 func (w Buffer) PrintableRuneWidth() int {
 	return PrintableRuneWidth(w.String())
 }
 
+// PrintableRuneWidth returns the cell width of the given string.
 func PrintableRuneWidth(s string) int {
 	var n int
 	var ansi bool
 
 	for _, c := range s {
-		if c == '\x1B' {
-			// ANSI escape sequence
-			ansi = true
-		} else if ansi {
-			if (c >= 0x40 && c <= 0x5a) || (c >= 0x61 && c <= 0x7a) {
-				// ANSI sequence terminated
-				ansi = false
-			}
-		} else {
-			n += runewidth.RuneWidth(c)
-		}
+		accPrintableRuneWidth(c, &n, &ansi)
 	}
 
 	return n
 }
+
+// Truncate truncates a given string at the given printable cell width, leaving
+// any ansi sequences intact.
+func Truncate(s string, w int, tail string) string {
+	if PrintableRuneWidth(s) <= w {
+		return s
+	}
+
+	var n int
+	var ansi bool
+	var acc strings.Builder
+
+	for _, c := range s {
+		accPrintableRuneWidth(c, &n, &ansi)
+		if n > w {
+			break
+		}
+
+		_, _ = acc.WriteRune(c)
+	}
+
+	if ansi {
+		// terminate any open ANSI sequences
+		_, _ = acc.WriteString("\x1B[0m")
```

Comment:
@kiyonlin thinking about this some more, I think we actually either always or need to send a reset sequence at the end. Consider something like:

s := Truncate("\x1B[7mHello", 1)
fmt.Println(s+"amburger")

In this case the entire string will print in reversed video because `\x1B[7m` is still active after the truncation. The question is, do we make it the responsibility of the user to handle that, or do we automatically reset ANSI after a truncate.

I could be wrong but I think the other modules in this package generally send reset sequences after operations.

Reply:
I think we should reset the state for the caller as it would otherwise just lead to weird styling issues.

Reply:
> I think we should reset the state for the caller as it would otherwise just lead to weird styling issues.

Agreed

Reply:
This is now in the branch.